src/NextGen/DonationForm/Controllers/DonateController.php

```diff
@@ -113,4 +120,38 @@ private function setSession(Donation $donation, Donor $donor)
             give_set_purchase_session($legacyPurchaseFormData->toPurchaseData());
         }
     }
+
+    /**
+     * @unreleased
+     *
+     * TODO: add more validation based on the field settings.
+     *
+     * @return void
+     */
+    private function saveCustomFields(DonationForm $form, Donation $donation, array $customFields)
+    {
+        $formSchema = $form->schema();
+
+        foreach ($customFields as $key => $value) {
+            /** @var Text $node */
+            $node = $formSchema->getNodeByName($key);
+
+            // make sure node exists in schema
+            if (!$node) {
+                throw new InvalidArgumentException("$key is not a valid custom field.");
+            }
+
+            if (empty($value) && $node->isRequired()) {
+                throw new InvalidArgumentException("$key is required.");
+            }
```

Comment:
@jonwaldstein What gets displayed to the user if this happens?

Reply:
We have a really basic and generic way of displaying server errors on the front-end so this would just display that message we have identified in the front-end.  Obviously this needs to be improved which goes back to our conversation about updating the front-end's interpretation of validation/errors.